### PR TITLE
Format/validate doc + synchronize code snippets with code

### LINKED ISSUE #132: Synchronize test-adapter and Getting Started Guide
Currently, the Getting Started Guide includes pieces of code which, combined, give a sample adapter, different from the test-adapter.

That leads to two issues:
- code in the Getting Started Guide is not compiled or tested. The last fix in #108 showed lots of inconsistencies (e.g. different name for a same variable in the code) or outdated code (e.g. the adapter had not been updated for the Interfaces changes).
- there's 2 "test adapters" to maintain, and both are meant to serve as an example implementation (the "test-adapter" is also used for unit tests and is compiled during CI). These test adapters may diverge.

To tackle the first issue, we could enforce the Markdown from Getting Started Guide to be synced with Go code (e.g. kube-prometheus uses [mdox](https://github.com/bwplotka/mdox) to check that Jsonnet examples from its Markdown documentation is synced with Jsonnet files - cf [here](https://github.com/prometheus-operator/kube-prometheus/blob/1fc4bc7e36732d555e8ed5662c086ce667b4eb20/docs/customizations/pod-anti-affinity.md?plain=1#L6)).

Ideally, we would have a single test implementation (test-adapter).
However, `test-adapter` is more complex than the implementation from the Getting Started Guide...

What do you think?

### PR DESCRIPTION
**Pull request by olivierlemasle**
This introduces [mdox](https://github.com/bwplotka/mdox) to:
- format / validate documentation markdown files
- synchronize snippets of code in Getting Started Guide with a new `docs/sample-adapter` adapter.

I'm not sure if it's a good thing to clutter `go.mod` with tooling dependencies such as `mdox` and `addlicense`; it may be better to use a separate module for tooling?

Fixes #132.

/kind documentation

### CODE REVIEW

**Comment by olivierlemasle** [comment]
Just discovered https://github.com/campoy/embedmd, which is used by kube-state-metrics. It might be easier...

File: hack/snippets/snippets.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
+// Copyright 2022 The Kubernetes Authors.
```
Comment (dgrisonnet) [inline hack/snippets/snippets.go:1]: is that coming from mdox?

Comment (olivierlemasle) [reply hack/snippets/snippets.go:1]: No, I wrote it :sweat:

**Review by dgrisonnet — COMMENTED** [review]
Maybe I am missing something but couldn't we use mdox-exec to add the go files to our doc? https://github.com/bwplotka/mdox#code-generation

**Comment by olivierlemasle** [comment]
@dgrisonnet Yes, that's what I'm using in this PR. For example:
 
    ```go mdox-exec="go run ./hack/snippets -d func=*yourProvider.ListAllMetrics docs/sample-adapter/pkg/provider/provider.go"

The objective of this `/hack/snippets` thing is just to print a part (here the function `*yourProvider.ListAllMetrics`) and not the whole file.

**Comment by dgrisonnet** [comment]
Maybe this could be contributed to mdox? I am not sure about how it would be integrated to the existing workflow though.

cc @bwplotka

**Comment by olivierlemasle** [comment]
I thought it was out of scope for mdox, but then I've just discovered https://github.com/bwplotka/mdox/issues/92. I'll discuss it there.

**Pull request closed without merging**